apps/cyberstorm-remix/cyberstorm/utils/dapperSingleton.ts

```diff
@@ -0,0 +1,100 @@
+import { DapperTs } from "@thunderstore/dapper-ts";
+import { type RequestConfig } from "@thunderstore/thunderstore-api";
+import { getSessionTools } from "../security/publicEnvVariables";
+import { getCachedRequestPromise } from "./requestCache";
+
+type ConfigFactory = () => RequestConfig;
+
+let currentConfigFactory: ConfigFactory | undefined;
+const requestScopedProxies = new WeakMap<Request, DapperTs>();
+
+function resolveConfigFactory(): ConfigFactory {
+  if (currentConfigFactory) {
+    return currentConfigFactory;
+  }
+
+  const tools = getSessionTools();
+  currentConfigFactory = () => tools.getConfig();
+  return currentConfigFactory;
+}
+
+function updateDapperConfig(factory: ConfigFactory) {
+  currentConfigFactory = factory;
+  if (typeof window !== "undefined" && window.Dapper) {
+    window.Dapper.config = () => factory();
+  }
+}
+
+export function initializeClientDapper(factory?: ConfigFactory) {
+  if (typeof window === "undefined") {
+    return;
+  }
+
+  if (factory) {
+    updateDapperConfig(factory);
+  } else if (!currentConfigFactory) {
+    resolveConfigFactory();
+  }
+
+  if (!window.Dapper) {
+    const resolvedFactory = resolveConfigFactory();
+    window.Dapper = new DapperTs(() => resolvedFactory());
+  }
+}
+
+export function getClientDapper(): DapperTs {
```

Comment:
Remind me again what we discussed about not doing this on SSR loaders? Was there some other reason beyond it likely being of little practical value?